File: TypeScript/0053-maximum-subarray.ts
// 2020-08-08 02:28:20

function maxSubArray(nums: number[]): number {
  let max = - Number.MAX_VALUE;
  let cur = 0;

  for (let i = 0; i < nums.length; i++) {
    if (cur <= 0) {
      cur = nums[i];
    } else {
      cur += nums[i];
    }

    max = Math.max(max, cur);
  }

  return max;
};

File: Java/0053_Maximum_Subarray.java
// 2020-08-07 18:26:37

class Solution {
    public int maxSubArray(int[] nums) {
        int max = Integer.MIN_VALUE, cur = 0;

        for (int i = 0; i < nums.length; i++) {
            cur = cur <= 0 ? nums[i] : (cur + nums[i]);
            max = Math.max(max, cur);
        }

        return max;
    }
}